### PR TITLE
Add optional argument --only-active-window to :session-save.

### PR DESCRIPTION
**Pull request by danfis**
The new optional argument --only-active-window makes :session-save to
save only the tabs in the currently active window.

### CODE REVIEW

**Comment by lahwaacz** [comment]
I think this does not make much sense without adjusting the `general.save-session` option too.

**Comment by danfis** [comment]
Maybe to add a new value only-active-window besides true/false?
I have set save-session to true which saves all windows and defined alias for calling :session-save with --only-active-window as a way to make a sort of bookmark containing several pages.

**Comment by lahwaacz** [comment]
Well, qutebrowser's sessions were designed to be per-instance rather than per-window, so changing that will be a lot of work, most likely more than changing couple of options. There are multiple issues open for this, e.g. #572 and #1096. If you'd like a dwb-like session handling, you might be interested in [this wrapper script](https://wiki.archlinux.org/index.php/Qutebrowser#dwb-like_session_handling) instead.

File: doc/help/commands.asciidoc [hunk 1]
```diff
@@ -735,6 +735,7 @@ Save a session.
 * +*-c*+, +*--current*+: Save the current session instead of the default.
 * +*-q*+, +*--quiet*+: Don't show confirmation message.
 * +*-f*+, +*--force*+: Force saving internal sessions (starting with an underline).
+* +*-o*+, +*--only-active-window*+: Saves only tabs of the currently active window.
```
Comment (The-Compiler) [inline doc/help/commands.asciidoc]: Docs are generated from the docstrings in the code by running `scripts/dev/src2asciidoc.py`. Please edit the docstring instead and then run that script (or revert those changes and I'll run it when merging).

File: qutebrowser/misc/sessions.py [hunk 2]
```diff
@@ -214,10 +214,18 @@ def _save_tab(self, tab, active):
...
         """Get a dict with data for all windows/tabs."""
         data = {'windows': []}
-        for win_id in objreg.window_registry:
+        winlist = objreg.window_registry
+        if only_active_window:
+            active_window = QApplication.instance().activeWindow()
+            winlist = [getattr(active_window, 'win_id', None)]
+            if winlist[0] is None:
+                # Fall back to all windows
+                winlist = objreg.window_registry
```
Comment (The-Compiler) [inline qutebrowser/misc/sessions.py]: You can get the active window in the command handler (`def session_save`) instead, by having a `win_id` argument and adding a `@cmdutils.argument('win_id', win_id=True)` decorator. Then simply pass something like `only_window=win_id` here.

File: qutebrowser/misc/sessions.py [hunk 3]
```diff
@@ -214,10 +214,18 @@ def _save_tab(self, tab, active):
             data['history'].append(item_data)
         return data
 
-    def _save_all(self):
+    def _save_all(self, only_active_window = False):
```
Comment (The-Compiler) [inline qutebrowser/misc/sessions.py]: This should probably be a keyword-only argument (here and elsewhere), i.e. `self, *, arg=False`. Also, no spaces around `=` for argument default values.

**Comment by lahwaacz** [comment]
> I don't think it makes sense to adjust the `save-session` option - `:wq` *closes* all windows anyways, so you wouldn't want that to only save one window but close all of them, I think.

I was thinking that the new `save-session` option would actually save each named window into a separate session when I run `:quit`. Otherwise I find running `:session-save --only-active-window <name>` manually and having to close the window to prevent duplicates in the default session rather awkward...

**Comment by danfis** [comment]
Hi, I have sorted out the documentation and changed the way the window-id is retrieved.
But I'm lost on tests, I really don't even know how to run them, let alone how to write them...

File: qutebrowser/misc/sessions.py [hunk 4]
```diff
@@ -255,7 +260,8 @@ def _get_session_name(self, name):
                     name = 'default'
         return name
 
-    def save(self, name, last_window=False, load_next_time=False):
+    def save(self, name, last_window=False, load_next_time=False,
+                   only_window=None):
```
Comment (The-Compiler) [inline qutebrowser/misc/sessions.py]: nitpick: Dedent this to line up with the `self`, not with `name`

**Comment by The-Compiler** [comment]
You can run the relevant tests by installing `python-tox` and optionally `xvfb` (to avoid windows popping up), and then run `tox -e py36 -- tests/end2end/features/test_sessions_bdd.py`. If your distribution has Python 3.5 or 3.4, use `py35` or `py34` respectively. See the [contribution docs](https://github.com/qutebrowser/qutebrowser/blob/master/CONTRIBUTING.asciidoc#running-specific-tests) for some more examples.

As for writing them, take a look at the existing ones in [sessions.feature](https://github.com/qutebrowser/qutebrowser/blob/master/tests/end2end/features/sessions.feature). An example is this:

```gherkin

  Scenario: Saving a simple session
    When I open data/hello.txt
    And I open data/title.html in a new tab
    Then the session should look like:
      windows:
        - active: true
          tabs:
            - history:
              - url: about:blank
              - active: true
                url: http://localhost:*/data/hello.txt
            - active: true
              history:
              - active: true
                url: http://localhost:*/data/title.html
                title: Test title
```

You could do something like this:

- open multiple windows
- Use `:session-save --only-active-window` 
- Run `:window-only` and `:tab-only` to get in a clean state again
- Load the saved session
- Make sure only two windows are open now, with the expected tabs.

Does that make things clearer?

**Comment by The-Compiler** [comment]
Thanks!

**Comment by sim590** [comment]
I have found a serious issue. If you have windows `A` and `B` with diffrent content each and you only want to save window `A`, you proceed to do so with `:session-save --only-active-window session-A`. It will succeed. You can then, try to `:session-load session-A` the session and you will notice the right window is loaded. If then you do `:wq`, we expect that both windows `A` and `B` will be saved and it is. However, upon reopening qutebrowser and try `:session-load session-A`, you will notice that the `session-A` has been corrupted. It seems like `:wq` saves on `session-A` which breaks the functionnalty.

**Comment by The-Compiler** [comment]
@sim590 This is intended, `:wq` saves to the last loaded session by default. If you want to change that behavior, you can do `:set general session-default-name default`

**Comment by sim590** [comment]
@The-Compiler Nice, it works! Thank you!

**Pull request merged**